Add heap usage tracking metric for DAG.

aschran opened the pull request:
Towards MystenLabs/sui#5213.

aschran committed:
Add heap usage tracking metric for DAG.

Towards #852.

huitseeker commented on consensus/src/consensus.rs:
```diff
@@ -325,6 +325,9 @@ where
                         .set(state.dag.len() as i64);
+                    self.metrics
+                        .dag_size_bytes
+                        .set((mysten_util_mem::malloc_size(&state.dag) + std::mem::size_of::<Dag>()) as i64);
```
What is the order of magnitude of the error we expect from the second argument of the sum?

aschran replied:
Probably quite small, but this is a fixed value so I'd hope that the call to size_of is inlined as a constant? Not sure how to check that though.

huitseeker replied:
We actually came up on this a while ago: https://github.com/MystenLabs/sui/pull/3920#discussion_r943290604

aschran replied:
In what way do you mean you came up on it? Clearly the heap allocated part needs to be measured (which is what that discussion seemed to be about), but what's the harm in also adding the stack allocated part to the measurement? Is there a nuance I'm missing here?

huitseeker replied:
Sorry about that, my comment above was misguided and not very constructive, I was more thinking out loud, and that led you on a wild goose chase, which I apologize for.

huitseeker reviewed (commented):
So we have a [micro-bench in the consensus](https://github.com/MystenLabs/narwhal/blob/main/consensus/benches/process_certificates.rs), which you can run with 
```bash
cargo-criterion  --bench process_certificates --features "benchmark, pprof"
```
The benchmark can also create flamegraphs, which should help in measuring where the CPU time goes along this processing, see [this blog post](https://www.jibbow.com/posts/criterion-flamegraphs/) for details.

That benchmark is based on `process_certificates`, so it exercises none of the logic that you add, which is based on receiving a message in the `run` loop of the consensus, which then calls `process_certificates`, followed by your metric update.

But I'm wondering if it would be possible to tweak the benchmark itself to call perhaps an internal auxiliary function in the measurement loop, which would compute
```rust
(mysten_util_mem::malloc_size(&state.dag) + std::mem::size_of::<Dag>()) as i64
```
and then measure the e2e impact of this on the overall cost of processing one certificate.

What I'm trying to figure out is the overall computational burden of this metric computation on the performance-sensitive consensus loop. If it's negligible, fine. If it's not, there are several ways to go about diminishing this burden, among which:
- sampling (updating the metric once in a few calls),
- memoizing (getting the latest value out of the metric and only adding the size of the elements we added to this add-only structure)

Deciding which to go with is essentially predicated on having a quantitative sense of this computational cost.

aschran commented:
> So we have a [micro-bench in the consensus](https://github.com/MystenLabs/narwhal/blob/main/consensus/benches/process_certificates.rs), which you can run with
> 
> ```shell
> cargo-criterion  --bench process_certificates --features "benchmark, pprof"
> ```
> 
> The benchmark can also create flamegraphs, which should help in measuring where the CPU time goes along this processing, see [this blog post](https://www.jibbow.com/posts/criterion-flamegraphs/) for details.
> 
> That benchmark is based on `process_certificates`, so it exercises none of the logic that you add, which is based on receiving a message in the `run` loop of the consensus, which then calls `process_certificates`, followed by your metric update.
> 
> But I'm wondering if it would be possible to tweak the benchmark itself to call perhaps an internal auxiliary function in the measurement loop, which would compute
> 
> ```rust
> (mysten_util_mem::malloc_size(&state.dag) + std::mem::size_of::<Dag>()) as i64
> ```
> 
> and then measure the e2e impact of this on the overall cost of processing one certificate.
> 
> What I'm trying to figure out is the overall computational burden of this metric computation on the performance-sensitive consensus loop. If it's negligible, fine. If it's not, there are several ways to go about diminishing this burden, among which:
> 
> * sampling (updating the metric once in a few calls),
> * memoizing (getting the latest value out of the metric and only adding the size of the elements we added to this add-only structure)
> 
> Deciding which to go with is essentially predicated on having a quantitative sense of this computational cost.

I added sampling for now. The benchmark showed enough of a performance drop that it seemed worthwhile. I think the right long term solution is to add support for a metric that computes its value at collection time, at that point we should switch this over. (I may work on that soon since it'd be useful for other purposes too.)

huitseeker reviewed (approved):
Thanks a crapton for running through the paces of benchmarking the perf impact! I really appreciate it, @aschran !
The sampling looks great, and so does your thinking on how to further progress on this.
